src/include/kdbio.h

```diff
@@ -0,0 +1,241 @@
+/**
+ * @file
+ *
+ * @brief
+ *
+ * @copyright BSD License (see LICENSE.md or https://www.libelektra.org)
+ */
+#ifndef KDB_IO_H_
+#define KDB_IO_H_
+
+struct ElektraIoInterface;
+
+struct ElektraIoFdOperation;
+
+/**
+ * Information about a file descriptor watched by IO-Binding
+ */
+typedef struct ElektraIoFdOperation
```

Comment:
If it is public we should avoid exposing structs because it would prohibit any extensions.

Reply:
I see. This would require some kind of accessor (getter and setter) function for every struct member, right?
I inteded the structs to be used directly, e.g (from src/bindings/io/uv/test/test_fd.c):
```C
ElektraIoFdOperation * fdOp = elektraIoTestSuiteUtilNewFdOperation ();  // malloc
fdOp->fd = fds[1];
fdOp->flags = ELEKTRA_IO_WRITABLE;
fdOp->enabled = 1;
fdOp->callback = testFdShouldSignalWritableProbe;

ElektraIoInterface * binding = createBinding (); // malloc + set pointers
binding->addFd (binding, fdOp);
```
But this means the code would transform into:
```C
// with a combined creator function (malloc + set struct members)
ElektraIoFdOperation * fdOp = elektraIoNewFdOperation (
  fds[1],  // int fd
  ELEKTRA_IO_WRITABLE,  // int flags
  1,  // int enabled
  NULL,  // void * privateData
  testFdShouldSignalWritableProbe  // callback
);

ElektraIoInterface * binding = createBinding ();
// when the ElektraIoInterface struct is not public, 
// this would also require functions to call binding functions
// instead of binding->addFd (binding, fdOp);
elektraIoBindingAddFd (binding, fdOp); 
```

I'm just asking because I'm not sure if that's what you meant; and it will take some work :)

I have to admit `elektraIoNewFdOperation()` is quite nice, because it will reduce possibilities for errors (e.g. forgot to fill a struct member).

Reply:
You got the idea. Get/set for every member is not always necessary, maybe your proposed constructor elektraIoNewFdOperation is enough. (Depending on if there are use cases which requires changes of already constructed FdOperation.) If getter/setter are required, maybe a single getter/setter is enough: one parameter could be used to tell which part should be changed, similar to the sysconf (3) call. In case of doubt (if something is needed), keep the API minimal.